src/adc.rs

```diff
@@ -0,0 +1,636 @@
+                pub fn read_linear_calibration_values(&mut self) -> AdcCalLinear {
+                    self.check_linear_read_conditions();
+
+                    // Read 1st block of linear correction
+                    self.rb.cr.modify(|_, w| w.lincalrdyw1().clear_bit());
+                    while self.rb.cr.read().lincalrdyw1().bit_is_set() {}
+                    let res_1 = self.rb.calfact2.read().lincalfact().bits();
+
+                    // Read 2nd block of linear correction
+                    self.rb.cr.modify(|_, w| w.lincalrdyw2().clear_bit());
+                    while self.rb.cr.read().lincalrdyw2().bit_is_set() {}
+                    let res_2 = self.rb.calfact2.read().lincalfact().bits();
+
+                    // Read 3rd block of linear correction
+                    self.rb.cr.modify(|_, w| w.lincalrdyw3().clear_bit());
+                    while self.rb.cr.read().lincalrdyw3().bit_is_set() {}
+                    let res_3 = self.rb.calfact2.read().lincalfact().bits();
+
+                    // Read 4th block of linear correction
+                    self.rb.cr.modify(|_, w| w.lincalrdyw4().clear_bit());
+                    while self.rb.cr.read().lincalrdyw4().bit_is_set() {}
+                    let res_4 = self.rb.calfact2.read().lincalfact().bits();
+
+                    // Read 5th block of linear correction
+                    self.rb.cr.modify(|_, w| w.lincalrdyw5().clear_bit());
+                    while self.rb.cr.read().lincalrdyw5().bit_is_set() {}
+                    let res_5 = self.rb.calfact2.read().lincalfact().bits();
+
+                    // Read 6th block of linear correction
+                    self.rb.cr.modify(|_, w| w.lincalrdyw6().clear_bit());
+                    while self.rb.cr.read().lincalrdyw6().bit_is_set() {}
+                    let res_6 = self.rb.calfact2.read().lincalfact().bits();
+
+                    AdcCalLinear([res_1, res_2, res_3, res_4, res_5, res_6])
+                }
+
+                fn check_linear_read_conditions(&self) {
+                    // Ensure the ADC is enabled and is not in deeppowerdown-mode
+                    if self.rb.cr.read().deeppwd().bit_is_set() {
+                        panic!("Cannot read linear calibration value when the ADC is in deeppowerdown-mode");
+                    }
+                    if self.rb.cr.read().advregen().bit_is_clear() {
+                        panic!("Cannot read linear calibration value when the voltage regulator is disabled");
+                    }
+                    if self.rb.cr.read().aden().bit_is_clear() {
+                        panic!("Cannot read linear calibration value when the ADC is disabled");
+                    }
+                }
+
+                fn reset(&mut self, ahb: &mut $AHB) {
+                    ahb.rstr().modify(|_, w| w.$adcxrst().set_bit());
+                    ahb.rstr().modify(|_, w| w.$adcxrst().clear_bit());
```

Comment:
The reset line `adc12en` is shared between ADC1 and ADC2. Therefore currently ADC1 and ADC2 can't be safely instantiated at the same time.
One solution would be to add a check in `init`, so that reset is only called if the AHB clock was previously disabled.